aws/resource_aws_efs_file_system_test.go

```diff
@@ -2,18 +2,95 @@ package aws
 
 import (
 	"fmt"
+	"log"
 	"regexp"
 	"testing"
+	"time"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/service/efs"
 
+	multierror "github.com/hashicorp/go-multierror"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
 	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/terraform"
 )
 
+func init() {
+	resource.AddTestSweepers("aws_efs_file_system", &resource.Sweeper{
+		Name: "aws_efs_file_system",
+		F:    testSweepEfsFileSystems,
+		Dependencies: []string{
+			"aws_efs_mount_target",
+		},
+	})
+}
+
+func testSweepEfsFileSystems(region string) error {
+	client, err := sharedClientForRegion(region)
+	if err != nil {
+		return fmt.Errorf("error getting client: %s", err)
+	}
+	conn := client.(*AWSClient).efsconn
+
+	err = testSweepIterateEfsFileSystems(conn, region, "EFS File Systems", func(id string) error {
+		log.Printf("[INFO] Deleting EFS File System: %s", id)
+
+		_, err := conn.DeleteFileSystem(&efs.DeleteFileSystemInput{
+			FileSystemId: aws.String(id),
+		})
+		if err != nil {
+			return fmt.Errorf("error deleting EFS File System %q: %w", id, err)
+		}
+
+		err = waitForDeleteEfsFileSystem(conn, id, 10*time.Minute)
+		if err != nil {
+			return fmt.Errorf("error waiting for EFS File System %q to delete: %w", id, err)
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+func testSweepIterateEfsFileSystems(conn *efs.EFS, region, resource string, iterator func(id string) error) error {
```

Comment:
Nit: This function feels a little complex and can be simplified with the AWS Go SDK paginated function ([`DescribeFileSystemsPages`](https://docs.aws.amazon.com/sdk-for-go/api/service/efs/#EFS.DescribeFileSystemsPages)). Outside of the paginated function existence though, It may be simpler in the future to just create a listing function then allow consumers to take the AWS Go SDK response output objects/error and do what they wish so its reusable across multiple code paths, e.g.

```go
func listEfsFileSystems(conn *efs.EFS) ([]*efs.FileSystem, error) {
	var fileSystems []*efs.FileSystem
	input := &efs.DescribeFileSystemsInput{}

	for {
		out, err := conn.DescribeFileSystems(input)
		if err != nil {
			return nil, err
		}

		fileSystems = append(fileSystems, out.FileSystems...)

		if out.NextMarker == nil {
			break
		}

		input.Marker = out.NextMarker
	}

	return fileSystems, err
}

// test sweeper implementation

fileSystems, err := listEfsFileSystems(conn)

if testSweepSkipSweepError(err) {
	log.Printf("[WARN] Skipping %s sweep for %s: %s", resource, region, err)
	return nil
}

if err != nil {
	return err
}

for _, fileSystem := range fileSystems {
  // do all the things with fileSystem.Id
}
```

Or we could make the implementation mimic the AWS Go SDK paginated functions, which have access to the whole response object for each page and would allow an easier replacement when the AWS Go SDK does support the paginated function.

In addition to the test sweepers, these types of "list" function primitives can be useful for plural data source implementations or singular data source implementations if we wanted support for a filter that wasn't available in the describe request parameters. 👍 See also the mount targets equivalent of this: https://github.com/terraform-providers/terraform-provider-aws/issues/2525

Reply:
I swear I searched for a paged function…